src/util/snippet.ts

```diff
@@ -0,0 +1,109 @@
+import { SnippetDefinition } from "../typings/snippet";
+import { Target } from "../typings/target.types";
+import {
+  Placeholder,
+  Text,
+  TextmateSnippet,
+  Variable,
+} from "../vendor/snippet/snippetParser";
+import { KnownSnippetVariableNames } from "../vendor/snippet/snippetVariables";
+
+/**
+ * Replaces the snippet variable with name `placeholderName` with
+ * TM_SELECTED_TEXT
+ *
+ * Also replaces any unknown variables with placeholders. We do this so it's
+ * easier to leave one of the placeholders blank. We may make it so that you can
+ * disable this with a setting in the future
+ * @param parsedSnippet The parsed textmate snippet to operate on
+ * @param placeholderName The variable name to replace with TM_SELECTED_TEXT
+ * @param substitutions A map from variable names to text values that will be
+ * substituted and the given variable will no longer be a placeholder in the
+ * final snippet
+ */
+export function transformSnippetVariables(
+  parsedSnippet: TextmateSnippet,
+  placeholderName?: string | null,
+  substitutions?: Record<string, string>
+): void {
+  let nextPlaceholderIndex = getMaxPlaceholderIndex(parsedSnippet) + 1;
+  const placeholderIndexMap: Record<string, number> = {};
+
+  parsedSnippet.walk((candidate) => {
+    if (candidate instanceof Variable) {
+      if (candidate.name === placeholderName) {
+        candidate.name = "TM_SELECTED_TEXT";
+      } else if (
+        substitutions != null &&
+        substitutions.hasOwnProperty(candidate.name)
+      ) {
+        candidate.parent.replace(candidate, [
+          new Text(substitutions[candidate.name]),
+        ]);
+      } else if (!KnownSnippetVariableNames[candidate.name]) {
+        let placeholderIndex: number;
+        if (candidate.name in placeholderIndexMap) {
+          placeholderIndex = placeholderIndexMap[candidate.name];
+        } else {
+          placeholderIndex = nextPlaceholderIndex++;
+          placeholderIndexMap[candidate.name] = placeholderIndex;
```

Comment:
This line should have been there all along; caused a bug with snippets that had a placeholder with the same name appearing multiple times